### ISSUE TITLE
Cannot rejoin public teams - public teams missing from join/select_team page

### ISSUE THREAD

**Issue by bbodenmiller** [mattermost/mattermost-server#9452]
#### Summary
After joining a public team then leaving it does not show up on select_team page for me to rejoin

#### Steps to reproduce
As Mattermost admin
1. Join public team
1. Leave public team
1. Clear cache and reload webpage
1. View join team page - notice that team just left is missing from join team page

Unknown if issue occurs for non-admins as well. On test could not reproduce. I also recently had done some work with `team add` CLI so unknown if that's connected or not.

#### Expected behavior
I should be able to join a public team regardless of if I was ever a member of the team or not

#### Observed behavior (that appears unintentional) 
Public teams I was previously a member of are missing from join/select_team page

#### Possible fixes
Unknown

**Comment by amyblais** [comment]
Hi @bbodenmiller, thank you for reaching out.

Do you happen to know if the team you saw this happen with was created via `team add` CLI?

I wasn't able to reproduce this, so this could be an issue with the CLI command.

**Comment by bbodenmiller** [comment]
I've tried adding my user using [`team add`](https://docs.mattermost.com/administration/command-line-tools.html#mattermost-team-add) then leaving the team and I cannot rejoin.

If join via team invite link and leave the team I see option to rejoin. As soon as I refresh the page option to rejoin disappears.

I'm also unable to duplicate on another server so I'm not really sure how to debug further.

**Comment by amyblais** [comment]
@bbodenmiller, Do you see any errors in logs or the Console when this issue happens? Also, can you help with any relevant details about your environment based on this [documentation](https://docs.mattermost.com/process/support.html#general-questions-for-any-issues)?

**Comment by amyblais** [comment]
Hi @bbodenmiller, we haven't received an update so we'll assume that the problem is fixed or is no longer valid. 

If you still experience the same problem, try upgrading to the [latest version](https://about.mattermost.com/download/). 

If the issue persists, reopen this issue with the relevant information and we'd be glad to help you where we can.

**Comment by bbodenmiller** [comment]
It doesn't look like I have permission to reopen but the issues back. I thought it had fixed itself but apparently not as a user just asked me to join their team and I couldn't find it. Tried accessing via direct link and it didn't work. Upon review from CLI the team exists and I could add myself via `team add`. Did a bit of verification and team is set to public. I then left the team via web UI and it still is not showing on join teams page.

I'm using Mattermost 5.3.1 as part of latest GitLab Omnibus release: 11.4.4

RHEL 7.3

Issue verified in Chrome and desktop app 4.1.1 on Windows

Here's a snippet from the logs (`sudo gitlab-ctl tail`... I trimmed a bunch of noise) when I ran `sudo -u mattermost /opt/gitlab/embedded/bin/mattermost --config=/var/opt/gitlab/mattermost/config.json team add <team> <user>`

File log level: ERROR

```
==> /var/log/gitlab/postgresql/current <==
2018-11-08_04:02:08.55409 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:02:08.55412 STATEMENT:  SELECT $1::regclass
2018-11-08_04:03:53.34016 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:03:53.34019 STATEMENT:  SELECT $1::regclass
2018-11-08_04:07:44.45567 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:07:44.45570 STATEMENT:  SELECT $1::regclass
2018-11-08_04:08:35.30734 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:08:35.30738 STATEMENT:  SELECT $1::regclass
2018-11-08_04:19:47.35761 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:19:47.35764 STATEMENT:  SELECT $1::regclass

==> /var/log/gitlab/mattermost/mattermost.log <==
{"level":"error","ts":1541650807.6452253,"caller":"app/team.go:491","msg":"Encountered an issue joining default channels err=SqlChannelStore.GetByName: Channel does not exist, teamId=[redact], name=off-topic, sql: no rows in result set","user_id":"[redact]","team_id":"[redact]"}
```

Now logs from `sudo -u mattermost /opt/gitlab/embedded/bin/mattermost --config=/var/opt/gitlab/mattermost/config.json team remove <team> <user>`

```
==> /var/log/gitlab/postgresql/current <==
2018-11-08_04:03:53.34016 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:03:53.34019 STATEMENT:  SELECT $1::regclass
2018-11-08_04:07:44.45567 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:07:44.45570 STATEMENT:  SELECT $1::regclass
2018-11-08_04:08:35.30734 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:08:35.30738 STATEMENT:  SELECT $1::regclass
2018-11-08_04:19:47.35761 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:19:47.35764 STATEMENT:  SELECT $1::regclass
2018-11-08_04:20:06.80842 ERROR:  relation "idx_teams_description" does not exist
2018-11-08_04:20:06.80845 STATEMENT:  SELECT $1::regclass
```

No other error messages referencing team left or user that left team. :/

**Comment by bbodenmiller** [comment]
Tried `sudo gitlab-ctl restart`, no luck.

**Comment by bbodenmiller** [comment]
Is there a way I can run the query that generates the data for the join team page to see if that shows it or not?

**Comment by amyblais** [comment]
This is Gitlab related so pinging @hmhealey 

Issue:
User cannot find/join open teams that he has previously left (either via team selection page or via a permalink).

Environment:
- Server 5.3.1 as part of latest GitLab Omnibus release: 11.4.4
- RHEL 7.3
- Chrome and desktop app 4.1.1 on Windows

Logs:
See https://github.com/mattermost/mattermost-server/issues/9452#issuecomment-436872398

Question:
Is there a way they can run the query that generates the data for the join team page to see if that shows it or not?

**Comment by hmhealey** [comment]
I'm not able to reproduce the issue with the given steps on https://community.mattermost.com running an RC for 5.5. I feel like there may have been a bug fixed around this recently relating to confusion around setting the team's type to open vs looking at the `Team.AllowOpenInvite` field, but I'm not able to find the ticket for it

**Comment by bbodenmiller** [comment]
Is there some database queries I can run to see if I'm getting unexpected
results? How is the list of teams to join constructed?

**Comment by hmhealey** [comment]
This will return all teams that should show up on that page
```sql
SELECT * FROM Teams WHERE AllowOpenInvite = true AND DeleteAt = 0;
```

**Comment by bbodenmiller** [comment]
But doesn't that show all teams, even the ones already part of so wouldn't but exact equivalent of join page?

**Comment by bbodenmiller** [comment]
Plot twist: I removed myself from being a system admin and now I'm seeing tons of new teams to join that weren't showing before, some which I had previously joined and left and others that I've never joined.

**Comment by hmhealey** [comment]
Hmm, that's very interesting. When you're a system admin, you *should* see all of the teams on the server that you aren't a member of, not just ones with `AllowOpenInvite` set. What's the total number of teams on the server, including deleted ones? It looks like there might be some discrepancy between the server and client handling on this since the server just does basically `SELECT * FROM Teams LIMIT 50` while the client is doing additional filtering based on whether or not the team has been deleted or actually whether `AllowOpenInvite` has been set (which it should ignore for system admins), so it's possible that the team you left and were trying to rejoin comes after those first 50 teams

**Comment by hmhealey** [comment]
Okay, so there's a few things causing this:
1. You're a system admin, so it's looking at [all teams](https://github.com/mattermost/mattermost-server/blob/master/store/sqlstore/team_store.go#L363) instead of [just open ones](https://github.com/mattermost/mattermost-server/blob/master/store/sqlstore/team_store.go#L423). This is intended.
2. It's only requesting [the first page of teams](https://github.com/mattermost/mattermost-webapp/blob/master/components/select_team/select_team.jsx#L54) from the server. This is kind of intended, but that screen could be improved to properly support the paging.
3. The results from the server even includes deleted teams and teams that you're a member of (see the links above), but the client [does additional filtering](https://github.com/mattermost/mattermost-redux/blob/master/src/selectors/entities/teams.js#L146), so not all of those are displayed. Also, it doesn't check if you're a system admin, so it filters out teams you should be able to join.

Long story short, you're not being sent that specific team from the server because you're receiving a bunch of teams that are not visible because the client thinks you can't join them. 

I've filed a series of tickets to:
1. Add paging to that screen on [mobile](https://mattermost.atlassian.net/browse/MM-13161) and [web](https://mattermost.atlassian.net/browse/MM-13160)
2. Make it so that system admins can see all teams that they can join (aka all teams) on that screen ([link](https://mattermost.atlassian.net/browse/MM-13163))
3. Fix it so that that endpoint doesn't return deleted teams ([link](https://mattermost.atlassian.net/browse/MM-13162))

**Comment by hanzei** [comment]
Closing this one, as 3 out of the 4 created ticket are closed.
